File: cpp/tests/unit-tests/ArbitrationResultTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2019 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include <memory>
#include <string>

#include <gtest/gtest.h>

#include "joynr/ArbitrationResult.h"

class ArbitrationResultTest: public ::testing::Test
{
public:
    ArbitrationResultTest()
        : arbitrationResult()
    {}

    joynr::ArbitrationResult arbitrationResult;
};

TEST_F(ArbitrationResultTest, testDiscoveryEntriesIsEmptySuccess)
{
    std::vector<joynr::types::DiscoveryEntryWithMetaInfo> discoveryEntries;
    arbitrationResult = joynr::ArbitrationResult(discoveryEntries);
    bool isArbitrationResultEmpty = arbitrationResult.isEmpty();
    EXPECT_TRUE(isArbitrationResultEmpty);
}

TEST_F(ArbitrationResultTest, testDiscoveryEntriesIsEmptyFailed)
{
    joynr::types::DiscoveryEntryWithMetaInfo discoveryEntry1 = joynr::types::DiscoveryEntryWithMetaInfo();
    discoveryEntry1.setParticipantId("participantId1");
    joynr::types::DiscoveryEntryWithMetaInfo discoveryEntry2 = joynr::types::DiscoveryEntryWithMetaInfo();
    discoveryEntry2.setParticipantId("participantId2");
    std::vector<joynr::types::DiscoveryEntryWithMetaInfo> expectedDiscoveryEntries {discoveryEntry2, discoveryEntry1};

    arbitrationResult = joynr::ArbitrationResult(expectedDiscoveryEntries);

    bool isArbitrationResultEmpty = arbitrationResult.isEmpty();
    EXPECT_FALSE(isArbitrationResultEmpty);
}

TEST_F(ArbitrationResultTest, testGetDiscoveryEntries)
{
    joynr::types::DiscoveryEntryWithMetaInfo discoveryEntry1 = joynr::types::DiscoveryEntryWithMetaInfo();
    discoveryEntry1.setParticipantId("participantId1");
    joynr::types::DiscoveryEntryWithMetaInfo discoveryEntry2 = joynr::types::DiscoveryEntryWithMetaInfo();
    discoveryEntry2.setParticipantId("participantId2");
    std::vector<joynr::types::DiscoveryEntryWithMetaInfo> expectedDiscoveryEntries {discoveryEntry2, discoveryEntry1};

    arbitrationResult = joynr::ArbitrationResult(expectedDiscoveryEntries);
    std::vector<joynr::types::DiscoveryEntryWithMetaInfo> actualDiscoveryEntries = arbitrationResult.getDiscoveryEntries();

    bool discoveryEntry1Found = false;
    bool discoveryEntry2Found = false;
    for (const auto& actualEntry : actualDiscoveryEntries) {
        if (actualEntry.getParticipantId() == discoveryEntry1.getParticipantId()) {
            discoveryEntry1Found = true;
        }
        if (actualEntry.getParticipantId() == discoveryEntry2.getParticipantId()) {
            discoveryEntry2Found = true;
        }
    }
    EXPECT_TRUE(discoveryEntry1Found && discoveryEntry2Found);
    EXPECT_EQ(expectedDiscoveryEntries.size(), actualDiscoveryEntries.size());
}

TEST_F(ArbitrationResultTest, testSetDiscoveryEntries)
{
    arbitrationResult = joynr::ArbitrationResult();
    std::vector<joynr::types::DiscoveryEntryWithMetaInfo> actualDiscoveryEntriesBeforeSet = arbitrationResult.getDiscoveryEntries();
    EXPECT_TRUE(actualDiscoveryEntriesBeforeSet.empty());

    joynr::types::DiscoveryEntryWithMetaInfo discoveryEntry1 = joynr::types::DiscoveryEntryWithMetaInfo();
    discoveryEntry1.setParticipantId("participantId1");
    joynr::types::DiscoveryEntryWithMetaInfo discoveryEntry2 = joynr::types::DiscoveryEntryWithMetaInfo();
    discoveryEntry2.setParticipantId("participantId2");
    std::vector<joynr::types::DiscoveryEntryWithMetaInfo> expectedDiscoveryEntriesAfterSet {discoveryEntry2, discoveryEntry1};
    arbitrationResult.setDiscoveryEntries(expectedDiscoveryEntriesAfterSet);
    std::vector<joynr::types::DiscoveryEntryWithMetaInfo> actualDiscoveryEntriesAfterSet = arbitrationResult.getDiscoveryEntries();

    bool discoveryEntry1Found = false;
    bool discoveryEntry2Found = false;
    for (const auto& actualEntry : actualDiscoveryEntriesAfterSet) {
        if (actualEntry.getParticipantId() == discoveryEntry1.getParticipantId()) {
            discoveryEntry1Found = true;
        }
        if (actualEntry.getParticipantId() == discoveryEntry2.getParticipantId()) {
            discoveryEntry2Found = true;
        }
    }
    EXPECT_TRUE(discoveryEntry1Found && discoveryEntry2Found);
    EXPECT_EQ(expectedDiscoveryEntriesAfterSet.size(), actualDiscoveryEntriesAfterSet.size());
}

File: java/core/libjoynr/src/test/java/io/joynr/arbitration/ArbitrationResultTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2021 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.arbitration;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import joynr.types.DiscoveryEntryWithMetaInfo;

public class ArbitrationResultTest {

    private ArbitrationResult arbitrationResult;

    @Before
    public void setUp() {
        arbitrationResult = new ArbitrationResult();
    }

    @Test
    public void initializeWithNonNullParameters() {
        DiscoveryEntryWithMetaInfo selectedDiscoverEntry = new DiscoveryEntryWithMetaInfo();
        DiscoveryEntryWithMetaInfo otherDiscoveryEntry1 = new DiscoveryEntryWithMetaInfo();
        DiscoveryEntryWithMetaInfo otherDiscoveryEntry2 = new DiscoveryEntryWithMetaInfo();

        Set<DiscoveryEntryWithMetaInfo> expectedSelectedDiscoveryEntries = new HashSet<DiscoveryEntryWithMetaInfo>(Arrays.asList(selectedDiscoverEntry));
        Set<DiscoveryEntryWithMetaInfo> expectedOtherDiscoveryEntries = new HashSet<DiscoveryEntryWithMetaInfo>(Arrays.asList(otherDiscoveryEntry1,
                                                                                                                              otherDiscoveryEntry2));
        arbitrationResult = new ArbitrationResult(expectedSelectedDiscoveryEntries, expectedOtherDiscoveryEntries);

        assertEquals(expectedSelectedDiscoveryEntries, arbitrationResult.getDiscoveryEntries());
        assertEquals(expectedOtherDiscoveryEntries, arbitrationResult.getOtherDiscoveryEntries());
    }

    @Test
    public void initializeWithNullParameters() {
        arbitrationResult = new ArbitrationResult(null, null);
        assertTrue(arbitrationResult.getDiscoveryEntries().isEmpty());
        assertTrue(arbitrationResult.getOtherDiscoveryEntries().isEmpty());
    }

    @Test
    public void setAndGetDiscoveryEntries() {
        Set<DiscoveryEntryWithMetaInfo> expectedSelectedDiscoveryEntries = new HashSet<DiscoveryEntryWithMetaInfo>(Arrays.asList(new DiscoveryEntryWithMetaInfo()));
        arbitrationResult.setDiscoveryEntries(expectedSelectedDiscoveryEntries);
        assertEquals(expectedSelectedDiscoveryEntries, arbitrationResult.getDiscoveryEntries());
    }

    @Test
    public void setDiscoveryEntriesWithNullParameters() {
        arbitrationResult.setDiscoveryEntries(null);
        assertTrue(arbitrationResult.getDiscoveryEntries().isEmpty());
    }

    @Test
    public void setAndGetOtherDiscoveryEntries() {
        Set<DiscoveryEntryWithMetaInfo> expectedOtherDiscoveryEntries = new HashSet<DiscoveryEntryWithMetaInfo>(Arrays.asList(new DiscoveryEntryWithMetaInfo()));
        arbitrationResult.setDiscoveryEntries(expectedOtherDiscoveryEntries);
        assertEquals(expectedOtherDiscoveryEntries, arbitrationResult.getDiscoveryEntries());
    }

    @Test
    public void setOtherDiscoveryEntriesWithNullParameters() {
        arbitrationResult.setOtherDiscoveryEntries(null);
        assertTrue(arbitrationResult.getOtherDiscoveryEntries().isEmpty());
    }
}
